crates/trie/trie/src/proof_v2/value.rs

```diff
@@ -0,0 +1,158 @@
+//! Generic value encoder types for proof calculation with lazy evaluation.
+
+use crate::{
+    hashed_cursor::HashedCursorFactory, proof_v2::ProofCalculator, trie_cursor::TrieCursorFactory,
+};
+use alloy_primitives::{B256, U256};
+use alloy_rlp::Encodable;
+use reth_execution_errors::trie::StateProofError;
+use reth_primitives_traits::Account;
+use reth_trie_common::RlpNode;
+use std::rc::Rc;
+
+/// A trait for deferred encoding of proof values.
+///
+/// This trait is implemented by types that can encode themselves into a buffer
+/// on demand, allowing for lazy computation of storage roots and other deferred work.
+pub trait ValueEncoderFut {
+    /// RLP encodes the value into the provided buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `buf` - A mutable buffer to encode the data into
+    fn encode(self, buf: &mut Vec<u8>) -> Result<(), StateProofError>;
+}
+
+/// A trait for encoding values for proof calculation.
+///
+/// This trait allows for lazy evaluation of encoding, enabling deferred
+/// computation of storage roots and other expensive operations.
+///
+/// The encoder takes a reference to itself and a value, returning a future-like
+/// type that will perform the encoding when needed.
+pub trait ValueEncoder {
+    /// The type of value being encoded (e.g., U256 for storage, Account for accounts).
+    type Value;
+
+    /// The type that will compute and encode the value when needed.
+    type Fut: ValueEncoderFut;
+
+    /// Returns a future-like value that will encode the value when called.
+    ///
+    /// # Arguments
+    ///
+    /// * `key` - The key the value was stored at in the DB
+    /// * `value` - The value to encode
+    ///
+    /// The returned future will be called as late as possible in the algorithm to maximize
+    /// the time available for parallel computation (e.g., storage root calculation).
+    fn encoder_fut(&self, key: B256, value: Self::Value) -> Self::Fut;
```

Comment:
I think we can also do `-> impl ValueEncoderFut` unless we need named types for this